File: apps/athenaeum/packages/domain/src/page-rpc.ts
import * as Schema from "effect/Schema"
import { EntityId } from "./node.js"
import { Page } from "./page.js"

// Wire schemas for `NotesService`'s page-body RPC methods (plan §"Storage & domain model":
// "Automerge note-body storage... Implement RPC methods to create a page (initializes an empty
// Automerge text doc), read a page's current text content, and apply a local edit (a plain
// text-replace op for Phase 1)"). Same one-Schema.Class-pair-per-method convention as rpc.ts/
// graph-rpc.ts.

export class CreatePageInput extends Schema.Class<CreatePageInput>("CreatePageInput")({
  workspaceId: EntityId,
  nodeId: EntityId
}) {}

export class CreatePageOutput extends Schema.Class<CreatePageOutput>("CreatePageOutput")({
  page: Page,
  text: Schema.String
}) {}

export class GetPageTextInput extends Schema.Class<GetPageTextInput>("GetPageTextInput")({
  workspaceId: EntityId,
  nodeId: EntityId
}) {}

export class GetPageTextOutput extends Schema.Class<GetPageTextOutput>("GetPageTextOutput")({
  page: Page,
  text: Schema.String
}) {}

/**
 * A plain text-replace op (plan: "a plain text-replace op for Phase 1 — a full ProseMirror
 * rich-text schema is NOT required this phase"): delete `deleteCount` UTF-16 code units starting
 * at `index`, then insert `insertText` at that same position — the same `(index, del, newText)`
 * shape `@automerge/automerge`'s own `splice` primitive takes, applied inside a single Automerge
 * `change` so it's one atomic local edit/one new CRDT change.
 */
export class ApplyPageEditInput extends Schema.Class<ApplyPageEditInput>("ApplyPageEditInput")({
  workspaceId: EntityId,
  nodeId: EntityId,
  index: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  deleteCount: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  insertText: Schema.String
}) {}

export class ApplyPageEditOutput extends Schema.Class<ApplyPageEditOutput>("ApplyPageEditOutput")({
  page: Page,
  text: Schema.String
}) {}

File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/PageRPC.swift
import Foundation

// Mirrors `packages/domain/src/page-rpc.ts` — `NotesService`'s page-body RPC methods:
// create a page (empty Automerge text doc), read current text, apply a local plain
// text-replace op.

public struct CreatePageInput: Codable, Hashable, Sendable {
    public let workspaceId: EntityId
    public let nodeId: EntityId
    public init(workspaceId: EntityId, nodeId: EntityId) {
        self.workspaceId = workspaceId
        self.nodeId = nodeId
    }
}

public struct CreatePageOutput: Codable, Hashable, Sendable {
    public let page: Page
    public let text: String
    public init(page: Page, text: String) {
        self.page = page
        self.text = text
    }
}

public struct GetPageTextInput: Codable, Hashable, Sendable {
    public let workspaceId: EntityId
    public let nodeId: EntityId
    public init(workspaceId: EntityId, nodeId: EntityId) {
        self.workspaceId = workspaceId
        self.nodeId = nodeId
    }
}

public struct GetPageTextOutput: Codable, Hashable, Sendable {
    public let page: Page
    public let text: String
    public init(page: Page, text: String) {
        self.page = page
        self.text = text
    }
}

/// Mirrors `page-rpc.ts`'s `ApplyPageEditInput` — a plain text-replace op: delete `deleteCount`
/// UTF-16 code units starting at `index`, then insert `insertText`. The same `(index, del,
/// newText)` shape `@automerge/automerge`'s own `splice` primitive takes.
public struct ApplyPageEditInput: Codable, Hashable, Sendable {
    public let workspaceId: EntityId
    public let nodeId: EntityId
    public let index: Int
    public let deleteCount: Int
    public let insertText: String

    public init(workspaceId: EntityId, nodeId: EntityId, index: Int, deleteCount: Int, insertText: String) {
        self.workspaceId = workspaceId
        self.nodeId = nodeId
        self.index = index
        self.deleteCount = deleteCount
        self.insertText = insertText
    }
}

public struct ApplyPageEditOutput: Codable, Hashable, Sendable {
    public let page: Page
    public let text: String
    public init(page: Page, text: String) {
        self.page = page
        self.text = text
    }
}
